src/server/serve.rs

```diff
@@ -53,25 +58,49 @@ pub enum PathType {
     SymlinkFile,
 }
 
+#[derive(Clone)]
+struct MakeService {
+    args: Arc<Args>,
+    gitignore: Arc<Gitignore>,
+}
+
+impl<T> Service<T> for MakeService {
+    type Response = InnerService;
+    type Error = hyper::Error;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+    fn poll_ready(&mut self, _: &mut Context) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, _: T) -> Self::Future {
+        let args = self.args.clone();
+        let gitignore = self.gitignore.clone();
+        let fut = async move { Ok(InnerService { args, gitignore }) };
+        Box::pin(fut)
+    }
+}
+
+impl MakeService {
+    pub fn new(args: Args) -> Self {
+        let args = Arc::new(args);
+        let gitignore = Arc::new(Gitignore::new(args.path.join(".gitignore")).0);
+        Self { args, gitignore }
+    }
+}
+
 /// Run the server.
 pub async fn serve(args: Args) -> BoxResult<()> {
     let address = args.address()?;
     let path_prefix = args.path_prefix.clone().unwrap_or_default();
-    let inner = Arc::new(InnerService::new(args));
-    let make_svc = make_service_fn(move |_| {
-        let inner = inner.clone();
-        async {
-            Ok::<_, Infallible>(service_fn(move |req| {
-                let inner = inner.clone();
-                inner.call(req)
-            }))
-        }
-    });
```

Comment:
I assumed that awaiting `handle_request` leads to performance loss, but the assumption seems incorrect after I ran some benchmarks.

484be4a `impl Service`

```
Running 10s test @ http://localhost:5000/Cargo.toml
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.09ms  813.62us  35.28ms   93.74%
    Req/Sec     4.91k   394.88     5.58k    79.50%
  97805 requests in 10.01s, 161.92MB read
Requests/sec:   9771.17
Transfer/sec:     16.18MB
```

481b1fb original functional `make_service_fn` invocation

```
Running 10s test @ http://localhost:5000/Cargo.toml
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     1.03ms  658.29us  15.17ms   84.55%
    Req/Sec     5.12k   493.27    10.27k    93.03%
  102418 requests in 10.10s, 163.80MB read
Requests/sec:  10140.85
Transfer/sec:     16.22MB
```

Performance even gets better from functional invocation! Actually it's `tokio::fs` leads to performance loss.

By the way, signature of `make_service_fn` is complicated, I struggled debugging with it. I'd like to implement `hyper::service::Service` on some structs e.g. `MakeService` and `InnerService` so I can concentrate on services instead of the whole `make_service_fn` invocation.

Since the above assumption is wrong and `Arc` has overhead on some platforms, I decide to revert to the original functional version on 481b1fb.